src/kernel/emulation/linux/misc/sysctl_machdep.c

```diff
@@ -51,3 +77,132 @@ sysctl_handler(handle_brand_string)
 	return 0;
 }
 
+sysctl_handler(handle_max_basic)
+{
+	setup(0);
+
+
+    __cpuid(level, eax, ebx, ecx, edx);
+
+
+	sprintf(old,"%d",eax);
+
+	return 0;
+}
+
+sysctl_handler(handle_family)
+{
+    setup(1);
+
+    __cpuid(level, eax, ebx, ecx, edx);
+
+    eax = eax >> 7;
+    eax &= 15;
+
+    sprintf(old,"%d",eax);
+
+    return 0;
+}
+
+sysctl_handler(handle_model)
+{
+    setup(1);
+
+    __cpuid(level, eax, ebx, ecx, edx);
+
+    eax = eax >> 3;
+    eax &= 15;
+
+    sprintf(old,"%d",eax);
+
+    return 0;
+}
+
+sysctl_handler(handle_stepping)
+{
+    setup(1);
+
+    __cpuid(level, eax, ebx, ecx, edx);
+
+    eax &= 15;
+
+    sprintf(old,"%d",eax);
+
+    return 0;
+}
+
+sysctl_handler(handle_brand_string)
+{
+    unsigned int level = 0;
+    unsigned int eax = 0x80000000;
+
+    unsigned int test[3];
+
+    __cpuid(level,eax,test, test+1,test+2);
+
+    if(eax < 0x80000004) // the information is not implemented
+        return 2;
+
+    eax = 0x80000001;
+
+    union
+    {
+        unsigned int brand[12];
+        char name[49];
+    } v;
+
+    __cpuid(level, eax, v.brand+0x1, v.brand+0x2, v.brand+0x3);
+    v.brand[0x0] = eax;
+    eax = 0x80000002;
+
+    __cpuid(level, eax, v.brand+0x5, v.brand+0x6, v.brand+0x7);
+    v.brand[0x4] = eax;
+    eax = 0x80000003;
+
+    __cpuid(level, eax, v.brand+0x9, v.brand+0xa, v.brand+0xb);
+    v.brand[0x8] = eax;
+
+    v.name[48] = 0;
+    copyout_string(v.name, (char*) old, oldlen);
+
+    return 0;
+}
+sysctl_handler(handle_features)
+{
+    setup(1);
+
+    char *features[] = {"FPU","VME", "DE", "PSE", "TSC", "MSR", "PAE", "MCE", "CX8", "APIC", NULL,"SEP","PGE","MCA",
+                        "CMOV", "PAT", "PSE-36", "PSN", "CLFSH", "DS", "ACPI", "MMX", "FXSR", "SSE", "SSE2", "SS",
+                        "HTT", "TM", NULL, "PBE"};
+
+    char *out = (char*)(calloc(119, sizeof(char))); //"worst" case: we have all of it plus null termination
```

Comment:
We actually don't have `calloc` in `src/kernel` (it's an undefined symbol at link time); use `malloc` instead (along with `memset`, if you really need to zero the memory).
```suggestion
    char *out = (char*)(malloc(119 * sizeof(char))); //"worst" case: we have all of it plus null termination
    memset(out, 0, 119 * sizeof(char));
```
Don't forget to also add the declarations for these functions to the top of the file:
```c
extern void* malloc(__SIZE_TYPE__ len);
extern void* memset(void* s, int c, size_t n);
```

Reply:
No, don't allocate memory on the heap at all. That's a weird thing to do for syscalls. Preallocate the memory, or allocate it on stack.

Reply:
Okay, but I'm not sure what do you mean by that @bugaevc it's my first time creating code for a kernel module.

Reply:
This is not the kernel module, that would be https://github.com/darlinghq/darling-newlkm 😀

This is userspace syscall emulation

Reply:
Basically I mean you cannot/shouldn't use `malloc/free`. Use `alloca`, or just static arrays.

Reply:
Okay and sorry. I thought it was the kernel module but after that, I realised that it's not because it's a separate repository. I'll use `alloca` then.